pkg/metallb/metallb.go

```diff
@@ -392,3 +393,101 @@ func (builder *Builder) convertToStructured(unsObject *unstructured.Unstructured
 
 	return metalLb, err
 }
+
+// WithMetallbControllerConfig defines the controller config with a helm chart table.
+func (builder *Builder) WithMetallbControllerConfig(priorityClassName, runtimeClassName, deploymentName,
+	helmChartControllerTestName, helmChartKeyExample string) *Builder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if !builder.Exists() {
+		glog.V(100).Infof("failed to update metallb, object doesn't exist on cluster")
+	}
+
+	glog.V(100).Infof("Defining the metallb object with Helm Chart in namespace %s",
+		builder.Definition.Namespace,
+	)
+
+	if builder.errorMsg != "" {
+		glog.V(100).Infof("Error occurred building metallb Controller Config")
+	}
+
+	builder.Definition.Spec.ControllerConfig = &mlbtypes.Config{PriorityClassName: priorityClassName,
+		RuntimeClassName: runtimeClassName, Annotations: map[string]string{deploymentName: "controller"}}
+
+	builder.Definition.Spec.ControllerConfig.Affinity = &corev1.Affinity{PodAffinity: &corev1.PodAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+			{
+				LabelSelector: &metaV1.LabelSelector{
+					MatchLabels: map[string]string{
+						"component": helmChartControllerTestName,
+					},
+				},
+				TopologyKey: "kubernetes.io/hostname",
+			},
+		},
+	},
+	}
+
+	builder.Definition.Spec.ControllerTolerations = []corev1.Toleration{
+		{
+			Key:      helmChartKeyExample,
+			Operator: "Exists",
+			Effect:   "NoExecute",
+		},
+	}
+
```

Comment:
We can define all 3( Spec, ControllerConfig, ControllerTolerations) together.

	builder.Definition.Spec = mlbtypes.MetalLBSpec{
		ControllerConfig: &mlbtypes.Config{
			PriorityClassName: priorityClassName,
			RuntimeClassName: runtimeClassName,
			Annotations: map[string]string{deploymentName: "controller"},
			Affinity: &corev1.Affinity{
				PodAffinity: &corev1.PodAffinity{
					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
						{
							LabelSelector: &metaV1.LabelSelector{
								MatchLabels: map[string]string{
									"component": helmChartControllerTestName,
								},
							},
							TopologyKey: "kubernetes.io/hostname",
						},
					},
				},
			},
		},
		ControllerTolerations: []corev1.Toleration{
			{
				Key:      helmChartKeyExample,
				Operator: "Exists",
				Effect:   "NoExecute",
			},
		},
	}

Reply:
I wanted to separate them to allow the configuration of a single element and to keep the function small.

Reply:
Good idea. Changed